### PR TITLE
`General`: Replace websocket connection alerts with connection lost indicator

### PR DESCRIPTION
**Pull request by bassner**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

If the websocket connection fails, the client currently displays a permanent alert, blocking access to the underlying components of the navbar.

image

### Description

This PR introduces a red indicator (wifi symbol) and a popover that belongs to it to notify the user about the connection interruption. The old notification alerts have been removed. This does not cover anything anymore, and shows the warning (in form of the animated symbol) even if the popover is clicked away.

As the new component needs access to the connection state, this PR additionally reworks how the websocket connection state is distributed in the application. Currently, the websocket service maintained arrays of connect and disconnect listeners, which are populated by calls to the ``bind()`` method of the service. This is not the best approach to this problem, and ignores the fact that we generally encourage the use of RxJs for usecases like this.

Therefore, the state of the websocket is now distributed using a RxJs behavior subject and comes with additional state information: connected, hasEverBeenConnected, intendedDisconnect.
All code that previously relied on connection events have been updated accordingly.

Lastly, I've updated the breakpoints and responsive mechanism of the navbar a bit as the new icon takes a bit more space (if it's rendered) and in general, currently, the navbar does not behave very well when it's scaled down (e.g. "Course Management" gets a line break etc). It will just switch to the collapsed navbar earlier than before. 

#### General
1. Start the web application
2. Disable your internet connection
--> Verify that the warning symbol in the top left pops up and that the popover is displayed.
--> You should be able to dismiss the popover and bring it back by clicking on the button again.
--> The pulse should stop after 1 minute; the icon itself stays there.
--> The icon should vanish if you reenable your internet connection.

#### Exam Mode
Please verify that the disconnect icon doesn't show the pulse animation *at all* if you are on the exam participation view as a student. Additionally, check that the connection indicator on the bottom when doing an exam as a student still shows up and updates if you disable your internet connection.

#### Navbar
Please scale down your viewport horizontally and observe that the navbar is responsive and looks good on smaller screen sizes, and that the hamburger menu still works.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|

### Screenshots

https://user-images.githubusercontent.com/23171488/160694426-cd238583-3b19-49df-bb55-f74502fbb359.mov

### CODE REVIEW

File: src/test/javascript/spec/component/exercises/quiz/quiz-participation.component.spec.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { ComponentFixture, discardPeriodicTasks, fakeAsync, TestBed, tick } from
 import { ActivatedRoute } from '@angular/router';
 import { NgbTooltip } from '@ng-bootstrap/ng-bootstrap';
 import { TranslateService } from '@ngx-translate/core';
-import { JhiWebsocketService } from 'app/core/websocket/websocket.service';
+import { ConnectionState, JhiWebsocketService } from 'app/core/websocket/websocket.service';
```
Comment (b-fein) [inline src/test/javascript/spec/component/exercises/quiz/quiz-participation.component.spec.ts:7]:
```suggestion
import { JhiWebsocketService } from 'app/core/websocket/websocket.service';
```
unused import

File: src/main/webapp/app/core/websocket/websocket.service.ts [hunk 2]
```diff
@@ -195,6 +207,10 @@ export class JhiWebsocketService implements IWebsocketService, OnDestroy {
         );
     }
 
+    private isConnected(): boolean {
+        return !!this.stompClient?.connected;
+    }
+
     /**
      * Close the connection to the websocket and unsubscribe als listeners.
```
Comment (b-fein) [inline src/main/webapp/app/core/websocket/websocket.service.ts:215]:
```suggestion
     * Close the connection to the websocket and unsubscribe all listeners.
```

**Review by b-fein — CHANGES_REQUESTED** [review]
Code looks good.

During testing the status got updated when disconnecting and reconnecting in the navbar, during quizzes, and during exams.
⟹ Everything worked flawlessly regarding the changes to the websocket connection. :+1:

Personal opinion: I’m not a fan of the pulsating icon and would rather like to have it without animation. I think that the red icon and the notification are noticeable enough.

### Requested Changes
The notifications now overflow the page to the right:
Screenshot 2022-03-30 at 13-42-05 Courses

The menu when clicking on the username is shown in the wrong location:
Screenshot 2022-03-30 at 13-42-36 Courses
The one for ‘Server Administration’ is shown correctly.

File: src/test/javascript/spec/component/exercises/quiz/quiz-participation.component.spec.ts (hunk 1 above)
Comment (bassner) [reply src/test/javascript/spec/component/exercises/quiz/quiz-participation.component.spec.ts]: 👍

**Comment by bassner** [comment]
@b-fein Thank you for testing! :)
The notification overflow should be fixed in 66b886c.
The dropdown location actually doesn't show up correctly on develop either, but with c68c4d0, it now works.

Regarding the animation topic: I've got the current appearance approved by @krusche, but I think wether or not we animate it is something we can quickly vote on in tomorrow's meeting. :)

**Review by b-fein — APPROVED** [review]
Thanks, all comments and issues I had before are now resolved.

**Comment by bassner** [comment]
Latest changes after today's dev meeting:
We keep the animation, but try to make it less distractive, especially if:
- The student is filling out an exam. The pulse animation should not be distracting from the exam.
- The websocket keeps disconnected for a longer period of time. The pulse doesn't need to keep blinking for a very long time, therefore we disable it after 1 minute.

All changes added in fd90d0a.

@b-fein I'd be super happy if you could verify these new changes again! :)

**Review by b-fein — DISMISSED** [review]
The icon stops the animation after a minute. On the exams page the icon does not play the animation at all. Code changes make sense. :+1: 

I like that the animation is disabled completely during an exam to avoid unnecessary distraction.

**Review by Santia-go — APPROVED** [review]
Tested on TS1. 
Working as expected in the general mode and the exam mode. 
Screenshot 2022-03-31 at 15 10 04

Additionally, the icon is responsive to the change of the scale in the window
Screenshot 2022-03-31 at 15 18 05

**Review by b-fein — APPROVED** [review]

**Comment by bassner** [comment]
@b-fein @Santia-go Just merged develop, nothing changed. :)

**Review by ole-ve — DISMISSED** [review]
Followed all testing steps on TS2, and worked as expected. 👍
Unfortunately, I cannot really comment on the (client) code.

**Review by Santia-go — APPROVED** [review]
Tested for a second time on TS1.
Working as expected on the two modes (general and exam).
The icon is adapted to the change of the scale in the window also.

**Review by mallgaier — DISMISSED** [review]
Good job, works as expected

**Pull request merged**